ticket_upgrade_prediction/models/neural_net_torch/train.py

```diff
@@ -0,0 +1,260 @@
+from torch.utils.data import DataLoader
+from tqdm import tqdm
+
+import mlflow
+from ticket_upgrade_prediction import Evaluator, Metrics
+from ticket_upgrade_prediction.config.env_config import EXPERIMENT_NAME
+from ticket_upgrade_prediction.pipeline import Dataset, Pipeline
+
+
+def mlflow_run_start_handle(method):
+    def wrapper(*args, **kwargs):
+        mlflow_run_name = kwargs.get("mlflow_run_name", None)
+        to_mlflow = True if mlflow_run_name else False
+
+        if to_mlflow:
+            client = mlflow.MlflowClient()
+            experiment = client.get_experiment_by_name(EXPERIMENT_NAME)
+
+            with mlflow.start_run(
+                run_name=mlflow_run_name,
+                experiment_id=experiment.experiment_id,
+            ):
+                method(*args, **kwargs)
+
+        else:
+            method(*args, **kwargs)
+
+    return wrapper
+
+
+@dataclass
+class HyperParams:
+    layers: list[int]
+    optimizer_name: str
+    learning_rate: float
+
+    def to_dict(self) -> dict:
+        return self.__dict__
+
+
+class NetworkTrainer:
+    def __init__(
+        self,
+        dataset: Dataset,
+        hparams: HyperParams = HyperParams(
+            layers=[5], optimizer_name="Adam", learning_rate=0.0001
+        ),
+        epochs: int = 2,
+        batch_size: int = 64,
+        criterion: Any = BCEWithLogitsLoss(),
+        plots_save_path: Optional[Path] = None,
+    ) -> None:
+        self.dataset = dataset
+        self.hparams = hparams
+        self.plot_save_path = plots_save_path
+        self.model = Network(
+            input_size=dataset.X_train.shape[1],
+            hidden_layers_sizes=self.hparams.layers,
+        )
+        self.optimizer = self._get_optimizer(
+            optimizer_name=self.hparams.optimizer_name,
+            learning_rate=self.hparams.learning_rate,
+        )
+        self.device = torch.device(
+            "cuda" if torch.cuda.is_available() else "cpu"
+        )
+        self.epochs = epochs
+        self.criterion = criterion
+        self.batch_size = batch_size
+
+        self.training_results = []
+
+    def _get_optimizer(self, optimizer_name: str, learning_rate: float) -> Any:
+        if optimizer_name.lower() == "adam":
+            return Adam(self.model.parameters(), lr=learning_rate)
+        elif optimizer_name.lower() == "sgd":
+            return SGD(self.model.parameters(), lr=learning_rate)
+        else:
+            raise ValueError("Wrong type of optimizer provided!")
+
+    def _one_batch(self, X: torch.Tensor, y: torch.Tensor) -> None:
+        X.to(self.device)
+        y.to(self.device)
+
+        preds_logit = self.model(X.float())
+        loss = self.criterion(preds_logit.float(), y.float())
+
+        self.optimizer.zero_grad()
+        loss.backward()
+        self.optimizer.step()
+
+    @mlflow_run_start_handle
+    def fit(self, mlflow_run_name: Optional[int] = None) -> None:
+        to_mlflow = True if mlflow_run_name else False
```

Comment:
i think this could be just to_mlflow = bool(mlflow_run_name)

Reply:
good point